### PR TITLE
clean up logic for easy understanding

### PR DESCRIPTION
**Pull request by jimengliu**
- Error out if extents are not ordered
- change the client side sync data logic to if statements, not loop
- modify a little bit both client and server binary search comments

@yasker @sheng-liang

### CODE REVIEW

File: sparse/sfold.go [hunk 1]
```diff
@@ -63,25 +63,33 @@ func coalesce(parentFileIo FileIoProcessor, childFileIo FileIoProcessor) error {
...
 
-		offset := data
-		buffer := AllocateAligned(blockSize)
-		for offset != hole {
-			// read a block from child, maybe use bufio or Reader stream
-			n, err := childFileIo.fileReadAt(buffer, offset)
-			if n != len(buffer) || err != nil {
+		batch := 32 * blockSize
+		if hole-data < int64(batch) {
+			batch = int((hole - data) * 32)
```
Comment (yasker) [inline sparse/sfold.go]: I don't understand what does this mean?

File: sparse/sfold.go [hunk 2]
```diff
@@ -63,25 +63,33 @@ func coalesce(parentFileIo FileIoProcessor, childFileIo FileIoProcessor) error {
 		}
 
 		// now we have a data start offset and length(hole - data)
-		// let's read from child and write to parent file block by block
+		// let's read from child and write to parent file. We read/write up to
```
Comment (yasker) [inline sparse/sfold.go]: It's not related to ssync change right? Please put sfold change in an separate PR.

Comment (jimengliu) [inline sparse/sfold.go]: yeah, it is not related to ssync. Sure I will generate another PR.

File: sparse/sfold.go (hunk 1 above)
Comment (jimengliu) [inline sparse/sfold.go]: Good catch. No one will understand it, including me :-( It is a typo(*32 part).  I should have double checked this before sending out, my bad.
